File: src/VMS.Modules.Trips/Models/FuelCardModels.cs
namespace VMS.Modules.Trips.Models;

public sealed class FuelCardModel
{
    public int FuelCardId { get; set; }
    /// <summary>§28: masked except the last 4 digits, e.g. <c>****1234</c>. The full number never leaves the server.</summary>
    public string MaskedCardNumber { get; set; } = string.Empty;
    public int FuelCardCompanyId { get; set; }
    public int? VehicleId { get; set; }
    public int? DriverId { get; set; }
    public string? CardHolderName { get; set; }
    public DateOnly ExpiryDate { get; set; }
    public decimal? MonthlyLimit { get; set; }
    public string Status { get; set; } = string.Empty;
    public string? Remarks { get; set; }
    public string RowVersion { get; set; } = string.Empty;
}

public sealed class CreateFuelCardRequest
{
    public string CardNumber { get; set; } = string.Empty;
    public int FuelCardCompanyId { get; set; }
    public string? CardHolderName { get; set; }
    public DateOnly ExpiryDate { get; set; }
    public decimal? MonthlyLimit { get; set; }
    public string? Remarks { get; set; }
}

/// <summary>The card number and its issuing company are set once, at creation, and not editable here — the same
/// "identifier, not a field" treatment this codebase gives a City's Abbreviation or a Currency's code.</summary>
public sealed class SaveFuelCardRequest
{
    public string? CardHolderName { get; set; }
    public DateOnly ExpiryDate { get; set; }
    public decimal? MonthlyLimit { get; set; }
    /// <summary>Restricted server-side to <c>FuelCardStatuses.Settable</c> — Expired cannot be set by hand.</summary>
    public string Status { get; set; } = string.Empty;
    public string? Remarks { get; set; }
    public string RowVersion { get; set; } = string.Empty;
}

public sealed class AssignFuelCardRequest
{
    public int? VehicleId { get; set; }
    public int? DriverId { get; set; }
    /// <summary>Defaults to today.</summary>
    public DateOnly? AssignedFrom { get; set; }
    public string? Reason { get; set; }
    public string RowVersion { get; set; } = string.Empty;
}

public sealed class FuelCardAssignmentModel
{
    public long FuelCardAssignmentId { get; set; }
    public int FuelCardId { get; set; }
    public int? VehicleId { get; set; }
    public int? DriverId { get; set; }
    public DateOnly AssignedFrom { get; set; }
    public DateOnly? AssignedTo { get; set; }
    public string? Reason { get; set; }
}

File: VMSFrontend/src/app/core/fuel-card.models.ts
/** Fuel Card Management (FSD §28, §48.3 screen 17). `fuelCardCompanyId` and `driverId` are Business Partner ids
 * (roles `FuelCardCompany` and `Driver`) — reuse `vms-partner-picker`, not a lookup. */
export type FuelCardStatus = 'Active' | 'Inactive' | 'Expired' | 'Blocked';

export interface FuelCardModel {
  fuelCardId: number;
  /** Masked except the last 4 digits (e.g. `****1234`) — the full number never leaves the server. */
  maskedCardNumber: string;
  fuelCardCompanyId: number;
  vehicleId?: number | null;
  driverId?: number | null;
  cardHolderName?: string | null;
  expiryDate: string;
  monthlyLimit?: number | null;
  status: FuelCardStatus;
  remarks?: string | null;
  rowVersion: string;
}

export interface CreateFuelCardRequest {
  cardNumber: string;
  fuelCardCompanyId: number;
  cardHolderName?: string | null;
  expiryDate: string;
  monthlyLimit?: number | null;
  remarks?: string | null;
}

/** The card number and its issuing company are set once, at creation, and not editable here. */
export interface SaveFuelCardRequest {
  cardHolderName?: string | null;
  expiryDate: string;
  /** Restricted server-side to Active/Inactive/Blocked — Expired is system-set only. */
  status: 'Active' | 'Inactive' | 'Blocked';
  monthlyLimit?: number | null;
  remarks?: string | null;
  rowVersion: string;
}

export interface AssignFuelCardRequest {
  vehicleId?: number | null;
  driverId?: number | null;
  /** Defaults to today. */
  assignedFrom?: string | null;
  reason?: string | null;
  rowVersion: string;
}

export interface FuelCardAssignmentModel {
  fuelCardAssignmentId: number;
  fuelCardId: number;
  vehicleId?: number | null;
  driverId?: number | null;
  assignedFrom: string;
  assignedTo?: string | null;
  reason?: string | null;
}
